rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,583 @@
+*   pyarrow.Schema
+    *   The stability of contents of the schema (i.e. how data is represented in
+        RecordBatches) is guaranteed by TFX-BSL, please see the “TFX-BSL”
+        section.
+
+For the entire TFX 1.x, at least PyArrow 1.x and 2.x will be supported.
+
+### TFX & ml-pipelines-sdk {#tfx-&-ml-pipelines-sdk}
+
+This section discusses all parts of the `tfx` package, including the minimal SDK
+package _ml-pipelines-sdk_, in detail.
+
+#### Artifact Payload {#artifact-payload}
+
+Almost all supported Artifact types have some payload on-disk storage. The
+compatibility guarantee is the following:
+
+*   Guaranteed Backward compatibility: within same major version, new release of
+    TFX components shall take Artifacts produced by an older version of TFX;
+    *   We shall record versions of all Artifacts: unless otherwise documented,
+        the version of an Artifact equals to the _&lt;major>.&lt;minor>_ version
+        of its producer component. This version will be further used as
+        consumption hint for these Artifacts, to correctly handle version
+        upgrades.
+*   **No Forward compatibility guarantee on payload**: while the system will not
+    reject passing Artifacts produced from newer versions of TFX to an old
+    version of a component, we do not guarantee this will work. A warning will
+    be generated in this case.
+*   Resolvers and importers would only resolve Artifacts produced with same or
+    older versions of the library, by default.
+
+#### SDK (aka DSL), Official Components {#sdk-aka-dsl-official-components}
+
+All parts of SDK (aka DSL), IR, interfaces of official components shall comply
+with SemVer. The functionalities form the following core CUJs, which include but
+are not limited to:
+
+*   Defining or declaring components;
+*   Composing them into a pipeline graph;
+*   Environment/Deployment specific decoration of nodes;
+*   Advanced control semantics (condition/loop).
+
+It’s worth noting again that APIs from underlying packages used in above core
+CUJs shall also comply.
+
+#### Custom Component {#custom-component}
+
+There are several ways to define a Custom Component in TFX::OSS:
+
+1.  Importing a user provided python function
+    *   this process and all APIs involved will be backward compatible;
+2.  Importing a user provided container image
+    *   this process and APIs will be backward compatible;
+3.  Defining ComponentSpec, ExecutorSpec and subclassing with component class
+    *   This is a very complex user interface, and arguably exposes too much
+        internal implementation details. Moreover, it’s proven that this is
+        quite difficult to do for our users. A possible improvement proposed is
+        to combine the component and executor classes, and have the convention
+        be that users subclass a base component's class to extend or override
+        functionality. Until this is settled, this API will not comply with
+        backward compatibility.
```

Comment:
@charlesccychen to see what we can try to finalize with current TFX 1.0 timeline.

Reply:
After some more thoughts, I think the `beam_pipeline_args` shouldn't be access from the custom component definition part of API anyway, here's my reasoning:

1. beam_pipeline_args are user input anyway, and a custom component also requires user input;
2. we are in the process of making this configurable to each component, and several non-beam based component like Trainer won't even see this flag after refactoring in https://github.com/tensorflow/tfx/commit/be20d084f2875655e6b94a908206ab8bf6ff278e

WDYT?

Reply:
This makes sense. Given https://github.com/tensorflow/tfx/pull/3584 I think it makes sense to not surface it anymore given beam_pipeline_args will be deprecated.

Reply:
@zhitaoli @casassg Regarding the beam args, it is crucial that the new logic for defining this on component level also supports fetching values from secrets. Otherwise the serialised component will contain sensitive args (e.g. s3 passwords) in clear text.

I have created a PR that solves this for the current version of TFX: https://github.com/tensorflow/tfx/pull/3455.

Reply:
> @zhitaoli @casassg Regarding the beam args, it is crucial that the new logic for defining this on component level also supports fetching values from secrets. Otherwise the serialised component will contain sensitive args (e.g. s3 passwords) in clear text.
> 
> I have created a PR that solves this for the current version of TFX: [tensorflow/tfx#3455](https://github.com/tensorflow/tfx/pull/3455).

I think it's technically decoupled from where the `beam_pipeline_args` are configured, pipeline or component.

Left a comment on tensorflow/tfx#3455 too for my concern about that direction.